packit/api.py

```diff
@@ -19,40 +19,51 @@ def __init__(self, config: Config, package_config: PackageConfig) -> None:
         self.config = config
         self.package_config = package_config
 
-    def sync_pr(self, pr_id, dist_git_branch: str, upstream_version: str = None):
-        up = Upstream(config=self.config, package_config=self.package_config)
+        self._up = None
+        self._dg = None
+
+    @property
+    def up(self):
+        if self._up is None:
+            self._up = Upstream(config=self.config, package_config=self.package_config)
+        return self._up
 
-        dg = DistGit(config=self.config, package_config=self.package_config)
+    @property
+    def dg(self):
+        if self._dg is None:
+            self._dg = DistGit(config=self.config, package_config=self.package_config)
+        return self._dg
```

Comment:
These lazy things make me a headache when working on the hooks/custom-scripts...;-)

(But no problem. Generally, it's nice to have it in this way...)

You can also use the `@lru_cache()` and avoid  defining the `self._property_name` at all... (But MYPY does not like it if I am not wrong.)

Reply:
I can try the @lru_cache

Reply:
You'll be sad, @lru_cache is just rubbish:
```python
    @property  # type: ignore
    @lru_cache()
    def up(self):
        return Upstream(config=self.config, package_config=self.package_config)
```
still gives sith ton of errors:
```
packit/api.py:34: error: "_lru_cache_wrapper[Any]" has no attribute "checkout_pr"
packit/api.py:37: error: "_lru_cache_wrapper[Any]" has no attribute "create_branch"
packit/api.py:42: error: "_lru_cache_wrapper[Any]" has no attribute "update_branch"                                                                                         
packit/api.py:43: error: "_lru_cache_wrapper[Any]" has no attribute "checkout_branch"                                                                                       
packit/api.py:45: error: "_lru_cache_wrapper[Any]" has no attribute "create_branch"                                                                                         
packit/api.py:46: error: "_lru_cache_wrapper[Any]" has no attribute "checkout_branch"
packit/api.py:48: error: "_lru_cache_wrapper[Any]" has no attribute "sync_files"                                                                                            
packit/api.py:48: error: "_lru_cache_wrapper[Any]" has no attribute "local_project"
packit/api.py:50: error: "_lru_cache_wrapper[Any]" has no attribute "create_patches"
packit/api.py:51: error: "_lru_cache_wrapper[Any]" has no attribute "local_project"
packit/api.py:53: error: "_lru_cache_wrapper[Any]" has no attribute "add_patches_to_specfile"                                                                               
packit/api.py:56: error: "_lru_cache_wrapper[Any]" has no attribute "local_project"                                                                                         
packit/api.py:60: error: Argument "distgit" to "sync" of "PackitAPI" has incompatible type "_lru_cache_wrapper[Any]"; expected "DistGit"
```

I give up.

Reply:
Grr, I am looking forward to [@cached_property](https://docs.python.org/dev/library/functools.html#functools.cached_property)...

> New in version 3.8.

..;-(

Reply:
:'(